ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFFromUnixTime.java

```diff
@@ -0,0 +1,173 @@
+    if (arguments.length < 1) {
+      throw new UDFArgumentLengthException("The function " + getName().toUpperCase() +
+          "requires at least one argument");
+    }
+    if (arguments.length > 2) {
+      throw new UDFArgumentLengthException("Too many arguments for the function " + getName().toUpperCase());
+    }
+    for (ObjectInspector argument : arguments) {
+      if (argument.getCategory() != Category.PRIMITIVE) {
+        throw new UDFArgumentException(getName().toUpperCase() +
+            " only takes primitive types, got " + argument.getTypeName());
+      }
+    }
+
+    PrimitiveObjectInspector arg0OI = (PrimitiveObjectInspector) arguments[0];
+    switch (arg0OI.getPrimitiveCategory()) {
+      case INT:
+        inputIntOI = (IntObjectInspector) arguments[0];
+        break;
+      case LONG:
+        inputLongOI = (LongObjectInspector) arguments[0];
+        break;
+      default:
+        throw new UDFArgumentException("The function " + getName().toUpperCase()
+            + " takes only int/long types for first argument. Got Type:" + arg0OI.getPrimitiveCategory().name());
+    }
+
+    if (arguments.length == 2) {
+      PrimitiveObjectInspector arg1OI = (PrimitiveObjectInspector) arguments[1];
+      switch (arg1OI.getPrimitiveCategory()) {
+        case CHAR:
+        case VARCHAR:
+        case STRING:
+          inputTextConverter = ObjectInspectorConverters.getConverter(arg1OI,
+              PrimitiveObjectInspectorFactory.javaStringObjectInspector);
+          break;
+        default:
+          throw new UDFArgumentException("The function " + getName().toUpperCase()
+              + " takes only string type for second argument. Got Type:" + arg1OI.getPrimitiveCategory().name());
+      }
+    }
+
+    if (timeZone == null) {
+      timeZone = SessionState.get().getConf().getLocalTimeZone();
+      formatter.setTimeZone(TimeZone.getTimeZone(timeZone));
+    }
+
+    return PrimitiveObjectInspectorFactory.writableStringObjectInspector;
+  }
+
+  public void configure(MapredContext context) {
+    if (context != null) {
+      String timeZoneStr = HiveConf.getVar(context.getJobConf(), HiveConf.ConfVars.HIVE_LOCAL_TIME_ZONE);
+      timeZone = TimestampTZUtil.parseTimeZone(timeZoneStr);
+      formatter.setTimeZone(TimeZone.getTimeZone(timeZone));
+    }
+  }
+
+  @Override
+  public Object evaluate(DeferredObject[] arguments) throws HiveException {
+    if (arguments[0].get() == null) {
+      return null;
+    }
+
+    if (inputTextConverter != null) {
+      if (arguments[1].get() == null) {
+        return null;
+      }
+      String format = (String) inputTextConverter.convert(arguments[1].get());
+      if (format == null) {
+        return null;
+      }
+      if (!format.equals(lastFormat)) {
+        formatter = new SimpleDateFormat(format);
+        formatter.setTimeZone(TimeZone.getTimeZone(timeZone));
+        lastFormat = format;
```

Comment:
This is coming from old udf. `format` is a parameter for the udf. Basically you try to avoid instantiating a new formatter for every row unless necessary (less overhead). That is why the last format is kept around.